### ISSUE TITLE
Equivalent of flax's TrainState

### ISSUE THREAD

**Issue by misterguick** [patrick-kidger/equinox#792]
Hi !

I'm quite new to jax in general and I'm moving from flax to equinox. I'm used to the TrainState feature they have. I haven't found something similar in equinox (yet). I'm trying to find what is the best _equinox_ way to achieve the same effect. I (ChatGPT)'ve quickly tested the following which seems to work but requires to use filtered transforms.

My questions are:

- Does it go against equinox design philosophy ?
- Is there a better way to achieve the same feature ?

Thank you very much !

EDIT: 

- We would probably not want to use flax's dataclass but rather pytrees ?
- **The example is MUCH slower when we do `step=self.step + 1` in the apply_gradients method. Do you have any idea why ?**

``` python
import optax
import jax
import equinox as eqx
from flax.struct import dataclass

@dataclass
class TrainState:
    model: eqx.Module
    tx: optax.GradientTransformation
    opt_state: optax.OptState
    step: int

    def apply_gradients(self, grads):
        updates, new_opt_state = self.tx.update(grads, self.opt_state, self.model)
        new_model = eqx.apply_updates(self.model, updates)
        return self.replace(model=new_model, opt_state=new_opt_state, step=self.step+1)

    @classmethod
    def create(cls, *, model, tx, **kwargs):
        opt_state = tx.init(model)
        return cls(model=model, tx=tx, opt_state=opt_state, step=0, **kwargs)
```

Example usage:

``` python
import jax
import jax.numpy as jnp
import equinox as eqx
import optax
from flax.struct import dataclass

class MLP(eqx.Module):
    layers: list

    def __init__(self, key):
        keys = jax.random.split(key, 3)
        self.layers = [
            eqx.nn.Linear(1, 64, key=keys[0]),
            eqx.nn.Linear(64, 64, key=keys[1]),
            eqx.nn.Linear(64, 1, key=keys[2]),
        ]

    def __call__(self, x):
        for layer in self.layers[:-1]:
            x = jax.nn.relu(layer(x))
        x = self.layers[-1](x)
        return x

import optax
import jax
import equinox as eqx
from flax.struct import dataclass

@dataclass
class TrainState:
    model: eqx.Module
    tx: optax.GradientTransformation
    opt_state: optax.OptState
    step: int

    def apply_gradients(self, grads):
        updates, new_opt_state = self.tx.update(grads, self.opt_state, self.model)
        new_model = eqx.apply_updates(self.model, updates)
        return self.replace(model=new_model, opt_state=new_opt_state, step=self.step + 1)

    @classmethod
    def create(cls, *, model, tx, **kwargs):
        opt_state = tx.init(model)
        return cls(model=model, tx=tx, opt_state=opt_state, step=0, **kwargs)

# Initialize the model
rng = jax.random.PRNGKey(0)
model = MLP(rng)

# Initialize the optimizer
tx = optax.adam(1e-3)
opt_state = tx.init(model)

# Create the train state
state = TrainState.create(model=model, tx=tx)

def compute_loss(model, x, y_true):
    y_pred = jax.vmap(model)(x)
    return jnp.mean((y_pred - y_true) ** 2)

@eqx.filter_jit
def train_step(state, x, y_true):
    def loss_fn(model):
        return compute_loss(model, x, y_true)

    grads = jax.grad(loss_fn)(state.model)
    state = state.apply_gradients(grads=grads)
    return state

# Generate training data
num_samples = 1000
x_train = jax.random.uniform(rng, (num_samples, 1), minval=0.0, maxval=10.0)
y_train = x_train  # The target is the same as the input

# Training loop
num_epochs = 10
batch_size = 32

for epoch in range(num_epochs):
    # Shuffle the data
    perm = jax.random.permutation(rng, num_samples)
    x_train = x_train[perm]
    y_train = y_train[perm]

    # Iterate over batches
    for i in range(0, num_samples, batch_size):
        x_batch = x_train[i:i + batch_size]
        y_batch = y_train[i:i + batch_size]
        state = train_step(state, x_batch, y_batch)

    # Compute the loss at the end of the epoch
    loss = compute_loss(state.model, x_train, y_train)
    print(f"Epoch {epoch + 1}, Loss: {loss}")

```

> Epoch 1, Loss: 0.6555840969085693
> Epoch 2, Loss: 0.022449418902397156
> Epoch 3, Loss: 0.0036858979146927595
> Epoch 4, Loss: 0.0006120167090557516
> Epoch 5, Loss: 0.00016658527601975948
> Epoch 6, Loss: 0.0001264694146811962
> Epoch 7, Loss: 9.197566396323964e-05
> Epoch 8, Loss: 7.977779750945047e-05
> Epoch 9, Loss: 6.888642383273691e-05
> Epoch 10, Loss: 6.138114258646965e-05

**Comment by patrick-kidger** [comment]
This looks reasonable to me!

The equivalent of `flax.struct.dataclass` would just be subclassing `eqx.Module`. Both of them are PyTree+dataclass combos. The differences are just that `eqx.Module` handles a lot of edge cases that Flax doesn't (inheritance, bound methods, abstract variables, ...).

FWIW I think it's less common to package things up into a 'train state' in Equinox -- take a look at [this introductory example as a reference](https://docs.kidger.site/equinox/examples/train_rnn/) -- but there's nothing wrong with the pattern.

On incrementing `step`: I think what's happening here is that you're actually recompiling on every call. `eqx.filter_jit` will trace all JAX arrays, and mark everything else as static. In this case `step` is an integer rather than an array. [`eqx.debug.assert_max_traces`](https://docs.kidger.site/equinox/api/debug/#equinox.debug.assert_max_traces) is a useful tool here.

**Comment by misterguick** [comment]
Thank you for your **super** quick reply !

Indeed you were right, setting the step field as a JAX array fixed the issue. I've changed dataclass to eqx.Module and played a bit with generics. Now it works as intended !

``` python
import jax
import jax.numpy as jnp
import equinox as eqx
import optax
from typing import TypeVar, Generic

# Define the MLP Model
class MLP(eqx.Module):
    layers: list[eqx.nn.Linear]

    def __init__(self, key: jax.random.PRNGKey):
        keys = jax.random.split(key, 3)
        self.layers = [
            eqx.nn.Linear(1, 64, key=keys[0]),
            eqx.nn.Linear(64, 64, key=keys[1]),
            eqx.nn.Linear(64, 1, key=keys[2]),
        ]

    def __call__(self, x: jnp.ndarray) -> jnp.ndarray:
        for layer in self.layers[:-1]:
            x = jax.nn.relu(layer(x))
        x = self.layers[-1](x)
        return x

# Define the TrainState with Step Field as a JAX array
M = TypeVar('M', bound=eqx.Module)

class TrainState(eqx.Module, Generic[M]):
    model: M
    tx: optax.GradientTransformation
    opt_state: optax.OptState
    step: jnp.ndarray

    def apply_gradients(self, grads: M):
        updates, new_opt_state = self.tx.update(grads, self.opt_state, self.model)
        new_model = eqx.apply_updates(self.model, updates)
        new_step = self.step + 1
        return eqx.tree_at(
            lambda state: (state.model, state.opt_state, state.step),
            self,
            (new_model, new_opt_state, new_step)
        )

    @classmethod
    def create(cls, model: M, tx: optax.GradientTransformation):
        opt_state = tx.init(model)
        return cls(model=model, tx=tx, opt_state=opt_state, step=jnp.array(0))

# Define the Loss Function and Training Step
def compute_loss(model: MLP, x: jnp.ndarray, y_true: jnp.ndarray) -> jnp.ndarray:
    y_pred = eqx.filter_vmap(model)(x)
    return jnp.mean((y_pred - y_true) ** 2)

@eqx.filter_jit
def train_step(state: TrainState[MLP], x: jnp.ndarray, y_true: jnp.ndarray) -> TrainState[MLP]:
    def loss_fn(model: MLP) -> jnp.ndarray:
        return compute_loss(model, x, y_true)

    grads = eqx.filter_grad(loss_fn)(state.model)
    return state.apply_gradients(grads)

# Initialize the model
rng = jax.random.PRNGKey(0)
model = MLP(rng)

# Initialize the optimizer
tx = optax.adam(1e-3)

# Create the train state using the new create method
state: TrainState[MLP] = TrainState.create(model=model, tx=tx)

# Generate training data
num_samples = 1000
x_train = jax.random.uniform(rng, (num_samples, 1), minval=0.0, maxval=10.0)
y_train = x_train  # The target is the same as the input

# Training loop
num_epochs = 10
batch_size = 32

for epoch in range(num_epochs):
    # Shuffle the data
    perm = jax.random.permutation(rng, num_samples)
    x_train = x_train[perm]
    y_train = y_train[perm]

    # Iterate over batches
    for i in range(0, num_samples, batch_size):
        x_batch = x_train[i:i + batch_size]
        y_batch = y_train[i:i + batch_size]
        state = train_step(state, x_batch, y_batch)

    # Compute the loss at the end of the epoch
    loss = compute_loss(state.model, x_train, y_train)
    print(f"Epoch {epoch + 1}, Loss: {loss}")
```
